app/src/core/player.ts

```diff
@@ -0,0 +1,56 @@
+import { DocumentGenerator, Paragraph } from './document';
+
+class Player {
+  sources: Record<string, HTMLVideoElement> = {};
+  playing = false;
+
+  pause(): void {
+    this.playing = false;
+    Object.values(this.sources).forEach((source) => {
+      source.pause();
+    });
+  }
+
+  async play(
+    paragraphs: Paragraph[],
+    time: number,
+    progressCallback: (time: number) => void
+  ): Promise<void> {
+    this.playing = true;
+    const renderItems = DocumentGenerator.fromParagraphs(paragraphs).from(time).toRenderItems();
+    for (const renderItem of renderItems) {
+      if (!this.playing) {
+        return;
+      }
+
+      const progress = () => {
```

Comment:
uuuhhh, hacky. but somehow i like it. are we sure that calling us again using `requestAnimationFrame` happens ofter enough? is there another way